README.md
## Splunk Connect for Kafka

Splunk Connect for Kafka is a Kafka Connect Sink for Splunk with the following features:

* Data ingestion from Kafka topics into Splunk via [Splunk HTTP Event Collector(HEC)](http://dev.splunk.com/view/event-collector/SP-CAAAE6M).
* In-flight data transformation and enrichment.

## Requirements
1. Kafka version 1.0.0 and above.
2. Java 8 and above.
3. A Splunk environment of version 6.5 and above, configured with valid HTTP Event Collector (HEC) tokens.

	* HEC token settings should be the same on all Splunk Indexers and Heavy Forwarders in your environment.
	* Task configuration parameters will vary depending on acknowledgement setting (See the [Configuration](#configuration) section for details).

	Note: HEC Acknowledgement prevents potential data loss but may slow down event ingestion.

## Build

1. Clone the repo from https://github.com/splunk/kafka-connect-splunk
2. Verify that Java8 JRE or JDK is installed.
3. Run `mvn package`. This will build the jar in the /target directory. The name will be "splunk-kafka-connect-v*.jar".

## Quick Start

1. [Start](https://kafka.apache.org/quickstart) your Kafka Cluster and confirm it is running.
2. If this is a new install, create a test topic (eg: `perf`). Inject events into the topic. This can be done using [Kafka data-gen-app](https://github.com/dtregonning/kafka-data-gen) or the Kafka bundle [kafka-console-producer](https://kafka.apache.org/quickstart#quickstart_send).
3. Place the jar file created by the maven build (splunk-kafka-connect-v*.jar) in or under the location specified in `plugin.path` (see step 4).
4. Within Kafka Connect, adjust values for `bootstrap.servers` and `plugin.path` inside `config/connect-distributed.properties` to fit your environment. The following values are also required:

    ```
	key.converter=<org.apache.kafka.connect.storage.StringConverter|org.apache.kafka.connect.json.JsonConverter|io.confluent.connect.avro.AvroConverter>
	value.converter=<org.apache.kafka.connect.storage.StringConverter|org.apache.kafka.connect.json.JsonConverter|io.confluent.connect.avro.AvroConverter>
	internal.key.converter=org.apache.kafka.connect.json.JsonConverter
	internal.value.converter=org.apache.kafka.connect.json.JsonConverter
	internal.key.converter.schemas.enable=false
	internal.value.converter.schemas.enable=false
	offset.flush.interval.ms=10000

	- For StringConverter and JsonConverter only -
	key.converter.schemas.enable=false
	value.converter.schemas.enable=false

	- For AvroConverter only -
	key.converter.schema.registry.url=<Location of Avro schema registry>
	value.converter.schema.registry.url=<Location of Avro schema registry>

    ```

5. Run `./bin/connect-distributed.sh config/connect-distributed.properties` to start Kafka Connect.
6. Run the following command to create connector tasks. Adjust `topics` to set the topic, and  `splunk.hec.token`  to set your HEC token.

    ```
	curl localhost:8083/connectors -X POST -H "Content-Type: application/json" -d '{
	"name": "kafka-connect-splunk",
	"config": {
	   "connector.class": "com.splunk.kafka.connect.SplunkSinkConnector",
	   "tasks.max": "3",
	   "topics":"<YOUR_TOPIC>",
	   "splunk.indexes": "",
	   "splunk.sources": "",
	   "splunk.sourcetypes": "",
	   "splunk.hec.uri": "https://localhost:8088",
	   "splunk.hec.token": "<YOUR_TOKEN>",
	   "splunk.hec.raw": "true",
	   "splunk.hec.raw.line.breaker": "",
	   "splunk.hec.json.event.enrichment": "<key value pairs separated by comma, only applicable to /event HEC>",
	   "splunk.hec.ack.enabled": "true",
	   "splunk.hec.ack.poll.interval": "10",
	   "splunk.hec.ack.poll.threads": "2",
	   "splunk.hec.ssl.validate.certs": "false",
	   "splunk.hec.http.keepalive": "true",
	   "splunk.hec.max.http.connection.per.channel": "4",
	   "splunk.hec.total.channels": "8",
	   "splunk.hec.max.batch.size": "1000000",
	   "splunk.hec.threads": "2",
	   "splunk.hec.event.timeout": "300",
	   "splunk.hec.socket.timeout": "120",
	   "splunk.hec.track.data": "true"
	  }
	}'
    ```

7. Verify that data is flowing into your Splunk platform instance by searching using the index, sourcetype or source from your configuration.
8. Use the following commands to check status, and manage connectors and tasks:

    ```
    # List active connectors
    curl http://localhost:8083/connectors

    # Get kafka-connect-splunk connector info
    curl http://localhost:8083/connectors/kafka-connect-splunk

    # Get kafka-connect-splunk connector config info
    curl http://localhost:8083/connectors/kafka-connect-splunk/config

    # Delete kafka-connect-splunk connector
    curl http://localhost:8083/connectors/kafka-connect-splunk -X DELETE

    # Get kafka-connect-splunk connector task info
    curl http://localhost:8083/connectors/kafka-connect-splunk/tasks
    ```

    See the [the Confluent doucumentation](https://docs.confluent.io/current/connect/managing.html#common-rest-examples) for additional REST examples.

See [Splunk Docs](http://docs.splunk.com/Documentation/KafkaConnect) for information on the following:

* Deployment
* Security
* Configuration
* Load balancing
* Benchmark Results
* Scaling out your environment
* Data loss and latency monitoring
* Troubleshooting


src/changes
From 588e947fe0e9584c62c8549207dae818d03f9281 Mon Sep 17 00:00:00 2001
From: Vitalii Rudenskyi <vitalii.rudenskyi@mckesson.com>
Date: Fri, 6 Jul 2018 11:25:07 -0700
Subject: [PATCH] * config style for topic metadata

---
 .gitignore                                         |   3 +-
 .../java/com/splunk/hecclient/JsonEventBatch.java  |  24 +
 .../java/com/splunk/hecclient/RawEventBatch.java   |  30 +-
 .../kafka/connect/SplunkSinkConnectorConfig.java   | 692 +++++++++++----------
 .../com/splunk/kafka/connect/SplunkSinkTask.java   | 142 +++--
 .../connect/SplunkSinkConnectorConfigTest.java     |  30 +-
 .../splunk/kafka/connect/SplunkSinkTaskTest.java   |  55 +-
 .../java/com/splunk/kafka/connect/UnitUtil.java    |   6 +-
 8 files changed, 593 insertions(+), 389 deletions(-)

diff --git a/.gitignore b/.gitignore
index d9ed333..f31ed44 100644
--- a/.gitignore
+++ b/.gitignore
@@ -27,4 +27,5 @@ target/*
 splunk-kafka-connect/
 pom.xml.versionsBackup
 .classpath
-.project
\ No newline at end of file
+.project
+/target/
diff --git a/src/main/java/com/splunk/hecclient/JsonEventBatch.java b/src/main/java/com/splunk/hecclient/JsonEventBatch.java
index 1f7f45a..43f81bd 100644
--- a/src/main/java/com/splunk/hecclient/JsonEventBatch.java
+++ b/src/main/java/com/splunk/hecclient/JsonEventBatch.java
@@ -15,6 +15,9 @@
  */
 package com.splunk.hecclient;

+import org.apache.commons.lang3.builder.EqualsBuilder;
+import org.apache.commons.lang3.builder.HashCodeBuilder;
+
 public final class JsonEventBatch extends EventBatch {
     public static final String endpoint = "/services/collector/event";
     public static final String contentType = "application/json; profile=urn:splunk:event:1.0; charset=utf-8";
@@ -43,4 +46,25 @@ public final class JsonEventBatch extends EventBatch {
     public EventBatch createFromThis() {
         return new JsonEventBatch();
     }
+
+
+    @Override
+    public int hashCode() {
+      return new HashCodeBuilder()
+          .append(endpoint)
+          .toHashCode();
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+      if (obj instanceof JsonEventBatch) {
+        final JsonEventBatch other = (JsonEventBatch) obj;
+        return new EqualsBuilder()
+            .append(endpoint, other.endpoint)
+            .isEquals();
+      } else {
+        return false;
+      }
+    }
+
 }
diff --git a/src/main/java/com/splunk/hecclient/RawEventBatch.java b/src/main/java/com/splunk/hecclient/RawEventBatch.java
index a5fbc47..905d25b 100644
--- a/src/main/java/com/splunk/hecclient/RawEventBatch.java
+++ b/src/main/java/com/splunk/hecclient/RawEventBatch.java
@@ -15,6 +15,8 @@
  */
 package com.splunk.hecclient;

+import org.apache.commons.lang3.builder.EqualsBuilder;
+import org.apache.commons.lang3.builder.HashCodeBuilder;
 import org.apache.http.client.utils.URIBuilder;

 public final class RawEventBatch extends EventBatch {
@@ -87,7 +89,7 @@ public final class RawEventBatch extends EventBatch {
             return this;
         }

-        public Builder setTime(final int time) {
+        public Builder setTime(final long time) {
             this.time = time;
             return this;
         }
@@ -146,4 +148,30 @@ public final class RawEventBatch extends EventBatch {
             params.addParameter(tag,  val);
         }
     }
+
+    @Override
+    public int hashCode() {
+      return new HashCodeBuilder()
+          .append(index)
+          .append(sourcetype)
+          .append(source)
+          .append(host)
+          .toHashCode();
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+      if (obj instanceof RawEventBatch) {
+        final RawEventBatch other = (RawEventBatch) obj;
+        return new EqualsBuilder()
+            .append(index, other.index)
+            .append(sourcetype, other.sourcetype)
+            .append(source, other.source)
+            .append(host, other.host)
+            .isEquals();
+      } else {
+        return false;
+      }
+    }
+
 }
diff --git a/src/main/java/com/splunk/kafka/connect/SplunkSinkConnectorConfig.java b/src/main/java/com/splunk/kafka/connect/SplunkSinkConnectorConfig.java
index 25d2c1f..10c95af 100644
--- a/src/main/java/com/splunk/kafka/connect/SplunkSinkConnectorConfig.java
+++ b/src/main/java/com/splunk/kafka/connect/SplunkSinkConnectorConfig.java
@@ -18,6 +18,7 @@ package com.splunk.kafka.connect;
 import com.splunk.hecclient.HecConfig;
 import org.apache.kafka.common.config.ConfigException;
 import org.apache.kafka.connect.sink.SinkConnector;
+import org.apache.kafka.connect.sink.SinkTask;
 import org.apache.kafka.common.config.AbstractConfig;
 import org.apache.kafka.common.config.ConfigDef;
 import org.apache.commons.lang3.StringUtils;
@@ -25,332 +26,395 @@ import org.apache.commons.lang3.StringUtils;
 import java.util.*;

 public final class SplunkSinkConnectorConfig extends AbstractConfig {
-    // General
-    static final String INDEX = "index";
-    static final String SOURCE = "source";
-    static final String SOURCETYPE = "sourcetype";
-    // Required Parameters
-    static final String URI_CONF = "splunk.hec.uri";
-    static final String TOKEN_CONF = "splunk.hec.token";
-    // General Parameters
-    static final String INDEX_CONF = "splunk.indexes";
-    static final String SOURCE_CONF = "splunk.sources";
-    static final String SOURCETYPE_CONF = "splunk.sourcetypes";
-    static final String TOTAL_HEC_CHANNEL_CONF = "splunk.hec.total.channels";
-    static final String MAX_HTTP_CONNECTION_PER_CHANNEL_CONF = "splunk.hec.max.http.connection.per.channel";
-    static final String MAX_BATCH_SIZE_CONF = "splunk.hec.max.batch.size"; // record count
-    static final String HTTP_KEEPALIVE_CONF = "splunk.hec.http.keepalive";
-    static final String HEC_THREDS_CONF = "splunk.hec.threads";
-    static final String SOCKET_TIMEOUT_CONF = "splunk.hec.socket.timeout"; // seconds
-    static final String SSL_VALIDATE_CERTIFICATES_CONF = "splunk.hec.ssl.validate.certs";

ci/perf.py
#!/usr/bin/python

import os
import time
import logging
import requests
import export_data

try:
    import urllib3
    urllib3.disable_warnings(urllib3.exceptions.InsecureRequestWarning)
except:
    pass

try:
    from requests.packages.urllib3.exceptions import InsecureRequestWarning
    requests.packages.urllib3.disable_warnings(InsecureRequestWarning)
except:
    pass

logging.basicConfig(
    format='%(asctime)-15s mod=%(module)s func=%(funcName)s line=%(lineno)d %(message)s',
    level=logging.INFO)

# envs
# KAFKA_CONNECT_HEC_MODE
# KAFKA_CONNECT_ACK_MODE
# KAFKA_CONNECT_TOPICS
# KAFKA_CONNECT_PERF_DURATION
# KAFKA_CONNECT_LINE_BREAKER
# INDEX_CLUSTER_SIZE
# JVM_HEAP_SIZE

CONNECTOR_URI = 'http://{}:8083/connectors'.format(
    os.environ.get('KAFKA_CONNECTOR_IP', 'kafkaconnect1'))
TOPIC = os.environ.get('KAFKA_CONNECT_TOPICS', 'perf')
JVM_HEAP_SIZE = os.environ.get('JVM_HEAP_SIZE', '8G')
LINE_BREAKER = os.environ.get('KAFKA_CONNECT_LINE_BREAKER', '@@@@')
TOKEN_WITHOUT_ACK = '00000000-0000-0000-0000-000000000000'
TOKEN_WITH_ACK = '00000000-0000-0000-0000-000000000001'
IDX_HOSTNAME_PREFIX = os.environ.get('IDX_HOSTNAME_PREFIX', 'idx')

PERF_CASES = [
    # {
    #     'tasks.max': '1',
    #     'splunk.hec.threads': '1',
    #     'splunk.hec.max.batch.size': '500',
    # },
    # {
    #     'tasks.max': '1',
    #     'splunk.hec.threads': '2',
    #     'splunk.hec.max.batch.size': '500',
    # },
    # {
    #     'tasks.max': '1',
    #     'splunk.hec.threads': '4',
    #     'splunk.hec.max.batch.size': '500',
    # },
    # {
    #     'tasks.max': '1',
    #     'splunk.hec.threads': '8',
    #     'splunk.hec.max.batch.size': '500',
    # },
    # {
    #     'tasks.max': '1',
    #     'splunk.hec.threads': '16',
    #     'splunk.hec.max.batch.size': '500',
    # },
    # {
    #     'tasks.max': '1',
    #     'splunk.hec.threads': '1',
    #     'splunk.hec.max.batch.size': '1000',
    # },
    # {
    #     'tasks.max': '1',
    #     'splunk.hec.threads': '2',
    #     'splunk.hec.max.batch.size': '1000',
    # },
    # {
    #     'tasks.max': '1',
    #     'splunk.hec.threads': '4',
    #     'splunk.hec.max.batch.size': '1000',
    # },
    # {
    #     'tasks.max': '1',
    #     'splunk.hec.threads': '8',
    #     'splunk.hec.max.batch.size': '1000',
    # },
    # {
    #     'tasks.max': '1',
    #     'splunk.hec.threads': '16',
    #     'splunk.hec.max.batch.size': '1000',
    # },
    # {
    #     'tasks.max': '2',
    #     'splunk.hec.threads': '1',
    #     'splunk.hec.max.batch.size': '500',
    # },
    # {
    #     'tasks.max': '4',
    #     'splunk.hec.threads': '1',
    #     'splunk.hec.max.batch.size': '500',
    # },
    # {
    #     'tasks.max': '8',
    #     'splunk.hec.threads': '1',
    #     'splunk.hec.max.batch.size': '500',
    # },
    # {
    #     'tasks.max': '16',
    #     'splunk.hec.threads': '1',
    #     'splunk.hec.max.batch.size': '500',
    # },
    {
        'tasks.max': '32',
        'splunk.hec.threads': '2',
        'splunk.hec.max.batch.size': '500',
    },
]


def get_hec_uris():
    # calculate HEC URIs
    indxer_cluster_size = int(os.environ['INDEX_CLUSTER_SIZE'])
    endpoints = []
    for i in xrange(1, indxer_cluster_size + 1):
        endpoints.append('https://{}{}:8088'.format(IDX_HOSTNAME_PREFIX, i))

    return ','.join(endpoints)


def _get_connector_config(hec_uris, hec_raw, hec_ack, test_case):
    token = TOKEN_WITHOUT_ACK
    if hec_ack == 'true':
        token = TOKEN_WITH_ACK

    hec_settings = 'raw_endpoint={raw}:use_ack={ack}'.format(
        raw=hec_raw, ack=hec_ack)

    params = ':'.join('{}={}'.format(k, v) for k, v in test_case.iteritems())
    sourcetype = 'connector-perf:{hec_settings}:{params}:jvm_heap={jvm}'.format(
        hec_settings=hec_settings, params=params, jvm=JVM_HEAP_SIZE)
    connector_name = 'splunk-sink-{}'.format(int(time.time() * 1000))

    connector_config = {
        'name': connector_name,
        'config': {
            'connector.class': 'com.splunk.kafka.connect.SplunkSinkConnector',
            'topics': TOPIC,
            'tasks.max': test_case['tasks.max'],
            'splunk.indexes': 'main',
            'splunk.sources': 'connector-perf',
            'splunk.sourcetypes': sourcetype,
            'splunk.hec.uri': hec_uris,
            'splunk.hec.token': token,
            'splunk.hec.ack.enabled': hec_ack,
            'splunk.hec.raw': hec_raw,
            'splunk.hec.max.batch.size': test_case['splunk.hec.max.batch.size'],
            'splunk.hec.threads': test_case['splunk.hec.threads'],
            'splunk.hec.track.data': 'true',
            'splunk.hec.ssl.validate.certs': 'false',
            'splunk.hec.raw.line.breaker': LINE_BREAKER,
            'name': connector_name,
        }
    }
    return connector_config


def create_connector(connector_uri, connector_config):
    logging.info('create connector %s', connector_config['name'])
    while True:
        try:
            resp = requests.post(connector_uri, json=connector_config)
        except Exception:
            logging.exception('failed to create connector')
            time.sleep(2)
            continue
        else:
            if resp.ok:
                return
            else:
                logging.error('failed to create connector %s', resp.json)
                time.sleep(2)


def delete_connector(connector_uri, connector_config):
    logging.info('delete connector %s', connector_config['name'])
    uri = '{}/{}'.format(connector_uri, connector_config['name'])
    while True:
        try:
            resp = requests.delete(uri)
        except Exception:
            logging.exception('failed to delete connector')
            time.sleep(2)
        else:
            if resp.ok:
                return
            else:
                logging.error('failed to delete connector %s', resp.text)
                if resp.status_code == 404:
                    return

                time.sleep(2)


def wait_for_connector_do_data_collection_injection():
    time.sleep(int(os.environ.get('KAFKA_CONNECT_PERF_DURATION', 3600)))


def _do_perf(hec_uris, hec_raw, hec_ack):
    metric_exporter = _new_data_exporter()
    start_time = time.time()
    for test_case in PERF_CASES:
        config = _get_connector_config(hec_uris, hec_raw, hec_ack, test_case)
        logging.info(
            'handling perf case: %s %s',
            config['config']['splunk.sources'],
            config['config']['splunk.sourcetypes'])
        create_connector(CONNECTOR_URI, config)
        wait_for_connector_do_data_collection_injection()
        delete_connector(CONNECTOR_URI, config)

        _export_metric(metric_exporter, start_time)
        start_time = time.time()


def _get_hec_configs():
    mode = os.environ.get('KAFKA_CONNECT_ACK_MODE')
    if mode == 'no_ack':
        ack_modes = ['false']
    elif mode == 'ack':
        ack_modes = ['true']
    else:
        ack_modes = ['true', 'false']

    # raw=true/false, ack=true/false
    hec_mode = os.environ.get('KAFKA_CONNECT_HEC_MODE')
    if hec_mode == 'event':
        hec_configs = {
            'false': ack_modes,
        }
    elif hec_mode == 'raw':
        hec_configs = {
            'true': ack_modes,
        }
    else:
        hec_configs = {
            'false': ack_modes,
            'true': ack_modes,
        }

    return hec_configs


def _new_data_exporter():
    config = export_data.ExportParams(
        src_splunk_uri='https://sh1:8089',
        src_splunk_user='admin',
        src_splunk_password='Chang3d!',
        dest_splunk_hec_uri='{}'.format(os.environ['CONNECT_PERF_METRIC_DEST_HEC']),
        dest_splunk_hec_token='{}'.format(os.environ['CONNECT_PERF_METRIC_TOKEN']),
        src_index='',
        src_sourcetypes=[],
        timeout=600,
    )
    return export_data.ExportData(config)


def _export_metric(exporter, start_time):
    end_time = time.time()
    queries = [
        'search index="_internal" source="*metrics*.log" group=per_sourcetype_thruput series="connector-perf*"',
        'search index="main" source="proc_monitor" sourcetype="metric"',
    ]
    for query in queries:
        for _ in xrange(3):
            try:
                exporter.export(query, start_time, end_time)
            except Exception:
                logging.exception(
                    'failed to export perf metrics, query=%s', query)
            else:
                break


def perf():
    hec_configs = _get_hec_configs()
    hec_uris = get_hec_uris()
    for hec_raw, hec_ack_enabled_settings in hec_configs.iteritems():
        for hec_ack_enabled in hec_ack_enabled_settings:
            _do_perf(hec_uris, hec_raw, hec_ack_enabled)


def _do_post(uri, data, auth):
    logging.info('posting to %s', uri)
    while 1:
        try:
            resp = requests.post(uri, data=data, auth=auth, verify=False)
        except Exception:
            logging.exception('failed to post to %s', uri)
            time.sleep(2)
        else:
            if resp.ok:
                return

            if resp.status_code == 409:
                # already exists
                return

            logging.error('failed to post to %s, error=%s', uri, resp.text)
            time.sleep(2)


def splunk_request(uri_data_gen):
    auth = requests.auth.HTTPBasicAuth(
        os.environ.get('SPLUNK_USER', 'admin'),
        os.environ.get('SPLUNK_PASS', 'Chang3d!'))

    indxer_cluster_size = int(os.environ['INDEX_CLUSTER_SIZE'])
    for i in xrange(1, indxer_cluster_size + 1):
        uri, data = uri_data_gen(i)
        _do_post(uri, data, auth)


def create_hec_token_with_ack():
    data = {
        'name': 'hec-token-ack',
        'token': TOKEN_WITH_ACK,
        'index': 'main',
        'indexes': 'main',
        'useACK': '1',
        'disabled': '0',
    }

    def uri_data_gen(i):
        uri = 'https://{}{}:8089/servicesNS/nobody/splunk_httpinput/data/inputs/http?output_mode=json'.format(IDX_HOSTNAME_PREFIX, i)
        return uri, data

    return splunk_request(uri_data_gen)


def config_line_breaker():
    data = {
        'name': 'source::connector-perf...',
        'LINE_BREAKER': '({})'.format(LINE_BREAKER),
        'SHOULD_LINEMERGE': 'false',
    }

    def uri_data_gen(i):
        uri = 'https://{}{}:8089/servicesNS/nobody/splunk_httpinput/configs/conf-props?output_mode=json'.format(IDX_HOSTNAME_PREFIX, i)
        return uri, data

    return splunk_request(uri_data_gen)


if __name__ == '__main__':
    create_hec_token_with_ack()
    config_line_breaker()
    perf()
